StRoot/StFcsDbMaker/StFcsDb.cxx

```diff
@@ -1979,3 +1979,125 @@ void StFcsDb::readGainCorrFromText(){
+			      g2ttrk[i].next_parent_p,g2ttrk[i].e)<<endm;
+  return i + 1;
+}
+    
+const g2t_track_st* StFcsDb::getParentG2tTrack(StFcsHit* h, g2t_track_st* g2ttrk, float& fraction, int& ntrk, unsigned int order){
+  const vector<pair<unsigned int, float>>& gt = h->getGeantTracks();  
+  ntrk=gt.size();
+  float detot=0;
+  if(order > gt.size()) {fraction=0; return 0;}
+  for(int itrk=0; itrk<ntrk; itrk++) detot += gt[itrk].second;
+  fraction = gt[order].second / detot;
+  if(mDebug>3){
+    for(unsigned int jtrk=0; jtrk<ntrk; jtrk++){
+      LOG_INFO << Form("Hit's parent G2T Track %3d id=%3d dE=%f",jtrk,gt[jtrk].first,gt[jtrk].second)<<endm;
+    }
+  }
+  return &g2ttrk[gt[order].first-1];
+}
+
+const g2t_track_st* StFcsDb::getPrimaryG2tTrack(StFcsHit* h, g2t_track_st* g2ttrk, float& fraction, int& ntrk,unsigned int order){
+  const vector<pair<unsigned int, float>>& gt = h->getGeantTracks();
+  vector<pair<unsigned int, float>> primary;
+  int n=gt.size();
+  float detot=0;
+  for(int itrk=0; itrk<n; itrk++){
+    float de=gt[itrk].second;
+    unsigned int id=backTraceG2tTrack(gt[itrk].first, g2ttrk);
+    int found=0;
+    unsigned int np=primary.size();
+    for(unsigned int jtrk=0; jtrk<np; jtrk++){
+      if(primary[jtrk].first == id) {primary[jtrk].second += de; found=1; break;}
+    }
+    if(found==0) primary.push_back(make_pair(id,de));
+    detot+=de;
+  }
+  ntrk = primary.size();
+  if(order > ntrk) {fraction=0; return 0;}
+  std::sort(primary.begin(), primary.end(),
+	    [](const pair<unsigned int,float>&a, const pair<unsigned int,float>&b){
+	      return b.second < a.second;
+	    });
+  fraction = primary[order].second / detot;
+  return &g2ttrk[primary[order].first-1];
+}
+
+const g2t_track_st* StFcsDb::getParentG2tTrack(StFcsCluster* c, g2t_track_st* g2ttrk, float& fraction, int& ntrk, unsigned int order){
+  float detot=0;
+  vector<pair<unsigned int, float>> parents;
+  StPtrVecFcsHit& hits = c->hits();  
+  int nhit = hits.size();
+  for(int ihit=0; ihit<nhit; ihit++){    
+    const vector<pair<unsigned int, float>>& gt = hits[ihit]->getGeantTracks();
+    int n=gt.size();
+    for(int itrk=0; itrk<n; itrk++){
+      unsigned int id=gt[itrk].first;
+      float de=gt[itrk].second;
+      int found=0;
+      unsigned int np=parents.size();
+      for(unsigned int jtrk=0; jtrk<np; jtrk++){
+	if(parents[jtrk].first == id) {parents[jtrk].second += de; found=1; break;}
+      }
+      if(found==0) parents.push_back(make_pair(id,de));
+      detot+=de;
+    }
+  }
+  ntrk=parents.size();
+  if(order > ntrk) {fraction=0; return 0;}
+  std::sort(parents.begin(), parents.end(),
+	    [](const pair<unsigned int,float>&a, const pair<unsigned int,float>&b){
+	      return b.second < a.second;
+	    });
+  fraction = parents[order].second / detot;
+  return &g2ttrk[parents[order].first-1];
+}
+
+const g2t_track_st* StFcsDb::getPrimaryG2tTrack(StFcsCluster* c, g2t_track_st* g2ttrk, float& fraction, int& ntrk, unsigned int order){
```

Comment:
Akio, you have quite a few common lines of code shared between the three new functions you introduce, i.e.  `getPrimaryG2tTrack(StFcsHit* h, ...)`, `getPrimaryG2tTrack(StFcsCluster* c, ...)`, and `getParentG2tTrack(StFcsCluster* c, .. )`. Actually, the latter two are identical except one single line and the difference between the first and second is that there is a loop over hits in the cluster. This usually indicates that there might be a better way to organize the code without repeating the logic.

Have you considered splitting the functionality into a few smaller more focused functions?
E.g. we could have a function to create a mapping between the Geant track id and its deposited energy for a cluster by looping over all hits in the cluster. It could be a method of `StFcsCluster` doing something like this:

```c++
TtoEMap accumulateHits() {
  TtoEMap cluster;
  auto addit = [](TtoEMap m, TtoEMap::value_type p) { ... };

  for (auto hit : hits) {
    cluster = accumulate(begin(hit.getGeantTracks()), end(hit.getGeantTracks()), cluster, addit);
  }

  return cluster;
}
```

where `TtoEMap` is the type of `mGeantTracks`

Then, there can be a function to re-map energies from the final state tracks to primary ones.

```c++
TtoEMap remapPrimary(TtoEMap m) {
  TtoEMap m_primary;
  for (auto e : m) {
    auto id = backTraceG2tTrack(e.first, g2ttrk)
    m_primary.get_or_create( id ) += e.second;
  }
  return m_primary;
}
```

Once you construct the maps, calculating `detot` and selecting the n-th highest element is trivial by using the std functions.

Reply:
Sorry for my ignorance. Maybe your pseudo code is not complete enough for me to understand. I guess I don't understand "TtoEMap is the type of mGeantTracks".

See what I did in the last push with getG2tTrack(StFcsCluster..., int mode=0) as a private function and calling it from getParentG2tTrack and getPrimaryG2tTrack with mode=0 & 1.

Reply:
> Maybe your pseudo code is not complete enough for me to understand. I guess I don't understand "TtoEMap is the type of mGeantTracks".

It means just what it says: `TtoEMap = delctype(StFcsHit::mGeantTracks)` which translates to `TtoEMap = vector<pair<unsigned, float>>` in your current code

> See what I did in the last push

Thanks. I think you will thank yourself later for that 😄  in case you need to make changes only in the private function rather than in a few places.

You can take it further from here and reuse `StFcsDb::getG2tTrack` in `StFcsDb::getParentG2tTrack(StFcsHit* h,` and `StFcsDb::getPrimaryG2tTrack(StFcsHit* h, ` by adding that single hit in a temporary "fake" cluster. I also see that there is a `StFcsCluster *StFcsHit::cluster()` method which perhaps can be used in case every hit is required to belong to a cluster.